NEW: Add new output for libnvptxcompiler

carterbox opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Separates new source feature from output splitting of #44 in order to make review of these concepts easier.

In this MR a new output is created for the libnvptxcompiler component.

`cuda-nvcc-impl` previously contained

```
lib/libnvptxcompiler_static.a
```
`cuda-nvcc-dev_linux-64` previously contained

```
targets/x86_64-linux/include/nvPTXCompiler.h
targets/x86_64-linux/lib/libnvptxcompiler_static.a 
```
These components are moved to the new libnvptxcompiler output.

The new package `libnvptxcompiler-dev_linux-64` now contains

```
lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a
targets/x86_64-linux/include/nvPTXCompiler.h
targets/x86_64-linux/lib/libnvptxcompiler_static.a
```

In order to ensure consistency between new and old package layouts we can show that the same artifacts are available in `cuda-nvcc-impl` by tracing the dependencies:

`cuda-nvcc-impl` depends on `cuda-nvcc-dev_linux-64` (noarch) which depends now on `libnvptxcompiler-dev_linux-64` (noarch), so these artifacts should still be accessible same as previously. However, the `cuda-nvcc-impl` package is now empty and contains no files directly.

> [!IMPORTANT]
> Because `libnvptxcompiler-dev_linux-64` (noarch) is now at the bottom of the dependency stack, `cuda-nvcc-dev_linux-64` (noarch) now contains a symbolic link `lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a`. this a problem? Will it cause clobbering between packages in the cross-compiling scenario?

carterbox committed:
REF: Create new output for libnvptxcompiler

Co-authored-by: Billy Suh <bsuh@nvidia.com>

carterbox committed:
MNT: Re-rendered with conda-build 25.5.0, conda-smithy 3.51.1, and conda-forge-pinning 2025.07.22.16.20.36

carterbox committed:
BLD: Bump the build number

carterbox committed:
REF: Remove direct dependency on new libnvptxcompiler package

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

carterbox committed:
REF: Move symlink back to cuda-nvcc-impl package

carterbox committed:
BUG: Fix compiler pinnings for Windows build

We want to preserve the vs2019 pinning since it seems that we test
this feedstock against the oldest supported compiler verison.

carterbox commented:
Because I believe that it is incorrect to have link `lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a` in a named-target package. I have moved this back to it's original package (`cuda-nvcc-impl`).

I have also fixed the broken Windows build because global pinnings recently dropped vs2019 for vs2022.

carterbox commented:
~~Why do we need / not need a `libnvptxcompiler-dev` package (no target suffix)?~~
https://github.com/conda-forge/cuda-nvcc-feedstock/pull/64

carterbox commented:
This link `lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a`

I guess it should go in `libnvptxcompiler-dev`, so why is that package implemented in the other feedstock? If we move that package here and add it to the dependencies of `cuda-nvcc-impl`, then the package split will be backwards compatible.

billysuh7 commented:
> This link `lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a`
> 
> I guess it should go in `libnvptxcompiler-dev`, so why is that package implemented in the other feedstock? If we move that package here and add it to the dependencies of `cuda-nvcc-impl`, then the package split will be backwards compatible.

It was originally done that way per suggestion from @jakirkham which I agreed seemed to make sense at the time. @jakirkham is there any concern if `libnvptxcompiler-dev` moved into this feedstock ?

billysuh7 commented:
> Because I believe that it is incorrect to have link `lib/libnvptxcompiler_static.a -> ../targets/x86_64-linux/lib/libnvptxcompiler_static.a` in a named-target package. I have moved this back to it's original package (`cuda-nvcc-impl`).
> 
> I have also fixed the broken Windows build because global pinnings recently dropped vs2019 for vs2022.

I think I get it, it makes sense to move the symlink out of the target-specific package in hindsight, thank you !

jakirkham commented:
> It was originally done that way per suggestion from @jakirkham which I agreed seemed to make sense at the time. @jakirkham is there any concern if libnvptxcompiler-dev moved into this feedstock ?

Thanks Daniel and Billy! 🙏

Yes this seems like a reasonable change to me

billysuh7 committed:
move lib/libnvptxcompiler_static.a symlink to libnvptxcompiler-dev

billysuh7 commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 25.5.0, conda-smithy 3.51.1, and conda-forge-pinning 2025.07.29.16.05.27

billysuh7 commented:
Hi @carterbox please have a look at my commit and let me know if I missed anything. The CI failures are because `libnvptxcompiler-dev` output was moved from the other NVCC feedstock. I need to somehow move this output using [admin-requests](https://github.com/conda-forge/admin-requests)

carterbox commented on recipe/meta.yaml:
```diff
@@ -146,6 +146,35 @@ outputs:
 
+  - name: libnvptxcompiler-dev
+    build:
+      noarch: generic
```
This package is not noarch. It is platform specific because it contains a link to a shared library, and it does not have a target_platform as part of the name.

```suggestion
```

billysuh7 replied:
> This package is not noarch. It is platform specific because it contains a link to a shared library, and it does not have a target_platform as part of the name.

Thank you for catching this! If there is nothing else, I guess I need to move `libnvptxcompiler-dev` out to this feedstock before merging. I will let you know.

carterbox committed:
REF: libnvptxcompiler-dev not noarch

billysuh7 commented:
@conda-forge-admin, please restart ci

billysuh7 commented:
CI is successful now

> Hi @carterbox please have a look at my commit and let me know if I missed anything. The CI failures are because `libnvptxcompiler-dev` output was moved from the other NVCC feedstock. I need to somehow move this output using [admin-requests](https://github.com/conda-forge/admin-requests)

CI is successful now after adding the needed output. Thanks.

carterbox commented:
LGTM!
